### PR TITLE
Qwen vl bounding box and overall vision fix

### PR DESCRIPTION
**Pull request by FMayran**
This PR adresses the root cause of issue #13694 which has been stale for quite a long time.
It is an alternate solution to https://github.com/ggml-org/llama.cpp/pull/15474 and is necessary because pull request #15474 causes incorrect positions to be used for positional encoding after an image.
The root issue is discussed in #13694 and https://github.com/ggml-org/llama.cpp/pull/15474 and comes down to this: for models with positional encoding, llamacpp expects token positions in the KV cache to be strictly increasing for causal masking, but Qwen VL and possibly others have a token pos that do not respect this (they can decrease for instance).

Although this PR does incur one root change inside ubatch (introducing a new field to store the position at which the token was inserted in the KV cache), positional encodings match transformer's implementation. The causal check uses this new position instead of the token pos.

This PR may be necessary to reach perfect accuracy in Qwen3VL, as mentionned in the discussion of #16207

This is my first PR.
Thanks @ggerganov @ngxson and especially @rujialiu

### CODE REVIEW

File: src/llama-batch.cpp [hunk 1]
```diff
@@ -705,21 +711,23 @@ llama_ubatch llama_batch_allocr::ubatch_add(const std::vector<int32_t> & idxs, u
...
+        /*.embd         =*/        batch.embd ? udata->embd.data() : nullptr,
+        /*.pos          =*/        udata->pos.data(),
+        /*.n_seq_id     =*/        udata->n_seq_id.data(),
+        /*.seq_id       =*/        udata->seq_id.data(),
+        /*.seq_id_unq   =*/        udata->seq_id_unq.data(),
+        /*.seq_idx      =*/        udata->seq_idx.data(),
+        /*.output       =*/        udata->output.data(),
+        /*.kv_position_of_token=*/ udata->kv_position_of_token.data(),
+        /*.data         =*/        std::move(udata),
+
```
Comment (pwilkin) [inline src/llama-batch.cpp:730]: Nitpick: reorder so that kv_position_of_token is last.

File: src/llama-kv-cache.cpp [hunk 2]
```diff
@@ -1215,6 +1216,12 @@ void llama_kv_cache::set_input_kq_mask(ggml_tensor * dst, const llama_ubatch * u
 
     std::fill(data, data + ggml_nelements(dst), -INFINITY);
 
+    std::vector<int32_t> map_kv_to_batch(n_kv, -1);//for each token in the cache, either (-1) or the position in the current ubatch
+    for (uint32_t i = 0; i < n_tokens; ++i)//invert the batch -> kv position map into a kv -> batch position map
+    {
+        if (ubatch->kv_position_of_token[i] != -1)
+            map_kv_to_batch[ubatch->kv_position_of_token[i]] = i;
+    }
```
Comment (pwilkin) [inline src/llama-kv-cache.cpp:1224]: You're filling a vector of int32_t's but iterating over uint32_t's, this is bound to trigger type conversion checks. Just iterate over int32_t.

**Review by pwilkin — CHANGES_REQUESTED** [review]

File: src/llama-batch.h [hunk 3]
```diff
@@ -49,6 +50,7 @@ struct llama_ubatch {
         std::vector<llama_seq_id>   seq_id_unq;
         std::vector<int32_t>        seq_idx;
         std::vector<int8_t>         output;
+        std::vector<int32_t>        kv_position_of_token;//when pushed to the kv cache, where is the token pushed (used for causal masking)
```
Comment (ngxson) [inline src/llama-batch.h:53]: I'd suggest a different naming for it, either `kv_pos` or `abs_pos` (absolute position). But it's up to @ggerganov to decide the naming here

File: tools/mtmd/mtmd.cpp [hunk 4]
```diff
@@ -1026,7 +1026,7 @@ const char * mtmd_image_tokens_get_id(const mtmd_image_tokens * image_tokens) {
 
 llama_pos mtmd_image_tokens_get_n_pos(const mtmd_image_tokens * image_tokens) {
     if (image_tokens->use_mrope_pos) {
-        return 1; // for M-RoPE, the whole image is 1 in temporal dimension
+        return (::std::max)(image_tokens->nx, image_tokens->ny);//assuming image, not video // for M-RoPE, the whole image is 1 in temporal dimension
```
Comment (ngxson) [inline tools/mtmd/mtmd.cpp:1029]:
```suggestion
        return std::max(image_tokens->nx, image_tokens->ny); // for M-RoPE, the image takes max(t, w, h) positions; t is omitted because we don't support video input
```

File: src/llama-batch.cpp [hunk 5]
```diff
@@ -256,36 +257,38 @@ bool llama_batch_allocr::init(
...
-        }
+        //@fmayran: these checks don't make sense with models using position encoding such as Qwen VL, because the position stored in the KV cache can jump around (it is not even always increasing).
+        //it is not enough to let them be repeating. Within an image embedding, arbitrary jumps are expected.
+        //const llama_pos p0 = memory ? memory->seq_pos_max(s) : -1;
+        //
+        //if (p0 >= 0) {
+        //    bool ok = true;
+        //
+        //    if (batch.token) {
+        //        if (seq_pos_min(s) != p0 + 1) {
```
Comment (ngxson) [inline src/llama-batch.cpp:268]: if this code block is safe to remove, better remove it altogether

**Review by ngxson — COMMENTED** [review]
Quick question: do we still need to provide the absolute (KV) position via `llama_batch`? I'm not sure what happen if the image is separated into multiple `llama_batch` before `llama_decode`

File: src/llama-batch.cpp (hunk 1 above)
Comment (FMayran) [reply src/llama-batch.cpp:730]: I am not sure about what you mean by that. It can't be after .data because it should be treated like other members of the update struct

File: src/llama-kv-cache.cpp (hunk 2 above)
Comment (FMayran) [reply src/llama-kv-cache.cpp:1224]: I will check to minimize casts, thanks for the reminder

File: tools/mtmd/mtmd.cpp (hunk 4 above)
Comment (FMayran) [reply tools/mtmd/mtmd.cpp:1029]: Regarding (::std::max), I am being extra careful because sometimes on windows max is a preprocessor definition and it breaks std::max. Parenthesis stop that behavior. If this is prevented for sure in llamacpp, we can remove them. And I will remove the useless :: scope

Comment (FMayran) [reply tools/mtmd/mtmd.cpp:1029]: Regarding the comment, it is not exactly "t", much like k.t with k a frame rate dependent constant and I will update the comment with the actual expression. I have some needs for video encoding in my side and may try and implement it in the future.

File: src/llama-batch.h (hunk 3 above)
Comment (FMayran) [reply src/llama-batch.h:53]: Yes, I was extra pedantic here :-)

File: src/llama-batch.cpp (hunk 1 above)
Comment (pwilkin) [reply src/llama-batch.cpp]: Sorry, you're right. This ordering is correct - but the one on line 212 is wrong (technically just misleading because it's just the comment, both structs are empty)

File: src/llama-batch.cpp [hunk 6]
```diff
@@ -224,6 +224,7 @@ bool llama_batch_allocr::init(
             /*.seq_idx      =*/ this->seq_idx.data(),
             /*.output       =*/ batch.logits,
             /*.data         =*/ {},
+            /*.kv_position_of_token=*/ {},
         };
```
Comment (pwilkin) [inline src/llama-batch.cpp]:
```suggestion
            /*.output       =*/ batch.logits,
            /*.kv_position_of_token=*/ {},
            /*.data         =*/ {},
        };
```

**Comment by FMayran** [comment]
> Quick question: do we still need to provide the absolute (KV) position via `llama_batch`? I'm not sure what happen if the image is separated into multiple `llama_batch` before `llama_decode`

From my understanding, the only use of the kv positions is causal making, which only needs to know if the considered token is after the current token. The batches should be inserted sequentially to the kV cache, so only the last batch matters and "after the current token in the current batch" means "after the current token". If apply_ubatch is not called on the last inserted ubatch, we do have a problem, yes, because my code only checks causality within the current batch.

File: src/llama-kv-cache.cpp [hunk 7]
```diff
@@ -895,6 +895,7 @@ void llama_kv_cache::apply_ubatch(const slot_info & sinfo, const llama_ubatch &
             }
 
             cells.pos_set(idx, ubatch.pos[i]);
+            ubatch.kv_position_of_token[i] = (int32_t)idx;//set the position in the kv cache as a property for this token (needed for proper causal masking)
```
Comment (ggerganov) [inline src/llama-kv-cache.cpp:898]: This does not work - we should never use the idx in the KV cache buffer for causality checks. Even if this works sometimes, in general nothing guarantees that the cache is ordered causally in memory. The simplest scenario in which this breaks is multiple sequences with unified KV cache.

**Review by ggerganov — COMMENTED** [review]
The PR in #15474 that I reviewed sometime ago seemed like a better solution. This PR makes an invalid assumption about the KV cell indices - namely that they can be used for causality checks.

**Comment by rujialiu** [comment]
> The PR in #15474 that I reviewed sometime ago seemed like a better solution. This PR makes an invalid assumption about the KV cell indices - namely that they can be used for causality checks.

So it looks like we really need, explicitly, another sequence 0,1,2,3,...  (as illustrated in your review) that's indepedent of KV cell indices and dedicated to causual masking.

**Comment by FMayran** [comment]
> > The PR in #15474 that I reviewed sometime ago seemed like a better solution. This PR makes an invalid assumption about the KV cell indices - namely that they can be used for causality checks.
> 
> So it looks like we really need, explicitly, another sequence 0,1,2,3,... (as illustrated in your review) that's indepedent of KV cell indices and dedicated to causual masking.

I see what you mean @ggerganov . The main issue is not actually to store the new sequence of positions, but rather the fact that most code assumes that the base value for positional encoding is equal to the token age in the cache. This in turn means @rujialiu 's https://github.com/ggml-org/llama.cpp/pull/15474 solution PLUS modifying mtmd_image_tokens_get_n_pos() to return both the kv cache sequence position and the positional encoding position, and updating everything that depends on these.
Or alternatively, we could add the handling of both positions in the implementation of the kv cache and add APIs to query both pos.

**Comment by rujialiu** [comment]
> I see what you mean @ggerganov . The main issue is not actually to store the new sequence of positions, but rather the fact that most code assumes that the base value for positional encoding is equal to the token age in the cache. This in turn means @rujialiu 's #15474 solution PLUS modifying mtmd_image_tokens_get_n_pos() to return both the kv cache sequence position and the positional encoding position, and updating everything that depends on these. Or alternatively, we could add the handling of both positions in the implementation of the kv cache and add APIs to query both pos.

Yes, that's why I didn't go further. I hoped to come up with a correct solution that has the following property:

1. The pure language models (and multi-modal models that don't "abuse" position_id) shouldn't incur any extra cost (time/memory)
2. The `pos` related variables should have consistent meanings.

However, 2 is difficult because, we need to go over lots of codes and think about "which `pos_id` is referred to". I suspect some of the codes would force us to add (and maintain) more variables than we initially planed.

And trying to keep 1 is even more difficult.

**Comment by FMayran** [comment]
@rujialiu I know a way to see the implications of such a change. We change the output of mtmd_image_tokens_get_n_pos() to an opaque struct and see what that breaks, and we start figuring things out by propagating the opaque struct and fixing things as we go.

**Comment by rujialiu** [comment]
> @rujialiu I know a way to see the implications of such a change. We change the output of mtmd_image_tokens_get_n_pos() to an opaque struct and see what that breaks, and we start figuring things out by propagating the opaque struct and fixing things as we go.

That's a brilliant idea!

File: tools/mtmd/mtmd.cpp (hunk 4 above)
Comment (ngxson) [reply tools/mtmd/mtmd.cpp:1029]: > I am being extra careful because sometimes on windows max is a preprocessor definition and it breaks std::max

`std::max` is used in multiple places across llama.cpp code base, there is nothing wrong with it. you should look to see how other files have certain `#define` to resolve the problem

**Comment by FMayran** [comment]
We can use an std::vector as the return type, it will break compilation all the same and it is probably the way to go to transmit as many "info on advancement/pos" as needed by future approaches, with the convention that [0] is the linearly increasing pos

**Comment by ngxson** [comment]
@FMayran I'll make an alternative fix based on your other PR https://github.com/ggml-org/llama.cpp/pull/15474 , I think it's a better solution overall. Leaving this message here to avoid duplicated efforts.

**Comment by FMayran** [comment]
Closed in favor of #16822

**Comment by FMayran** [comment]
> @FMayran I'll make an alternative fix based on your other PR #15474 , I think it's a better solution overall. Leaving this message here to avoid duplicated efforts.

Yes, I closed in favor of your implementation.
#15474 is @rujialiu 's work btw
